### ISSUE TITLE
Way to use tornado's static file handler?

### ISSUE THREAD

**Issue by tkhyn** [GaretJax/sphinx-autobuild#71]
Hi and thanks for sphinx-autobuild, 

I have issues retrieving large-ish JS files through tornado that's used behind the scenes by livereload. 

My documentation requires mathjax, which files are downloaded and served locally through sphinx-autobuild. When the documentation tries to retrieve MathMenu.js (which is 32.8 kb), I get the following error:

```
[E 181214 08:51:53 web:1670] Uncaught exception GET /_static/mathjax/extensions/MathMenu.js?V=2.7.5 (127.0.0.1)
    HTTPServerRequest(protocol='http', host='127.0.0.1:8008', method='GET', uri='/_static/mathjax/extensions/MathMenu.js?V=2.7.5', version='HTTP/1.1', remote_ip='127.0.0.1')
    Traceback (most recent call last):
      File "/home/thomas/.buildout/eggs/tornado-5.1.1-py3.5-linux-x86_64.egg/tornado/web.py", line 1592, in _execute
        result = yield result
      File "/home/thomas/.buildout/eggs/tornado-5.1.1-py3.5-linux-x86_64.egg/tornado/gen.py", line 1133, in run
        value = future.result()
      File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
        raise self._exception
      File "/home/thomas/.buildout/eggs/tornado-5.1.1-py3.5-linux-x86_64.egg/tornado/gen.py", line 326, in wrapper
        yielded = next(result)
      File "/home/thomas/.buildout/eggs/tornado-5.1.1-py3.5-linux-x86_64.egg/tornado/web.py", line 2528, in get
        yield self.flush()
      File "/home/thomas/.buildout/eggs/tornado-5.1.1-py3.5-linux-x86_64.egg/tornado/web.py", line 994, in flush
        start_line, self._headers, chunk, callback=callback)
      File "/home/thomas/.buildout/eggs/tornado-5.1.1-py3.5-linux-x86_64.egg/tornado/http1connection.py", line 412, in write_headers
        data += self._format_chunk(chunk)
      File "/home/thomas/.buildout/eggs/tornado-5.1.1-py3.5-linux-x86_64.egg/tornado/http1connection.py", line 424, in _format_chunk
        "Tried to write more data than Content-Length")
    tornado.httputil.HTTPOutputError: Tried to write more data than Content-Length
```

I'm opening this issue here because as far as I understand tornado can use a static files handler to deal with this kind of problem, and I think it would be good if we could tell the tornado server that it should use that handler for anything in the sphinx static folder.

I haven't looked too much at how livereload and tornado work how we could achieve it so opening the discussion here. From what I saw some modifications to livereload may be needed to access this functionality - you may know that better than I do!

**Comment by pradyunsg** [comment]
I don't! Maybe @GaretJax does.

**Comment by GaretJax** [comment]
Tornado should be more than capable of handling files of 30kB of size. It looks like a different issue caused by a mismatch between the computed file size and the actual data in the file. If I have to guess, maybe an encoding issue around this location: https://github.com/tornadoweb/tornado/blob/master/tornado/web.py#L2598

What version of tornado and Python is this happening on (if it is still happening, 2 years later)? Has probably been fixed upstream, and would anyway not be an issue with sphinx-autobuild.

**Comment by pradyunsg** [comment]
<img width="1313" alt="Screenshot 2020-08-27 at 12 40 06 AM" src="https://user-images.githubusercontent.com/3275593/91346021-e3a1df00-e7fd-11ea-82ee-6238756732f1.png">

I'm unable to reproduce the issue you're reporting.

**Comment by tkhyn** [comment]
Hi and thanks for looking into it,

It looks like only Mathjax.js gets loaded and to trigger requests for other files you need to add a mathjax formula to your `index.rst`, such as:

```
:math:`x + 2`
```

**Comment by pradyunsg** [comment]
Got it!

```
[E 200827 13:05:11 web:1788] Uncaught exception GET /_static/mathjax/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.7 (127.0.0.1)
    HTTPServerRequest(protocol='http', host='127.0.0.1:8000', method='GET', uri='/_static/mathjax/config/TeX-AMS-MML_HTMLorMML.js?V=2.7.7', version='HTTP/1.1', remote_ip='127.0.0.1')
    Traceback (most recent call last):
      File "/Users/pradyunsg/Projects/sphinx-autobuild/.nox/docs-live/lib/python3.8/site-packages/tornado/web.py", line 1703, in _execute
        result = await result
      File "/Users/pradyunsg/Projects/sphinx-autobuild/.nox/docs-live/lib/python3.8/site-packages/tornado/web.py", line 2653, in get
        await self.flush()
      File "/Users/pradyunsg/Projects/sphinx-autobuild/.nox/docs-live/lib/python3.8/site-packages/tornado/web.py", line 1101, in flush
        return self.request.connection.write(chunk)
      File "/Users/pradyunsg/Projects/sphinx-autobuild/.nox/docs-live/lib/python3.8/site-packages/tornado/http1connection.py", line 497, in write
        self._pending_write = self.stream.write(self._format_chunk(chunk))
      File "/Users/pradyunsg/Projects/sphinx-autobuild/.nox/docs-live/lib/python3.8/site-packages/tornado/http1connection.py", line 473, in _format_chunk
        raise httputil.HTTPOutputError(
    tornado.httputil.HTTPOutputError: Tried to write more data than Content-Length
[E 200827 13:05:11 web:1198] Cannot send error response after headers written
```

Download and extract https://github.com/mathjax/MathJax/archive/2.7.7.zip in _static/mathjax + use this conf.py:

```py
extensions = ["sphinx.ext.mathjax"]
project = "sphinx-autobuild"
html_static_path = ["_static"]
mathjax_path = "mathjax/MathJax.js?config=TeX-AMS-MML_HTMLorMML"
```

Commenting out mathjax_path makes things work.

**Comment by pradyunsg** [comment]
https://github.com/mkdocs/mkdocs/issues/1351 -- mkdocs hits the same issue.
https://github.com/tornadoweb/tornado/issues/2743 seems related.

https://github.com/lepture/python-livereload/issues/174 -- pretty sure this is an issue with vanilla python-livereload.

**Comment by pradyunsg** [comment]
```
[E 200827 13:33:37 handlers:192] Could not open static file '/Users/pradyunsg/Projects/sphinx-autobuild/build/docs/_static/mathjax/extensions/MathMenu.js'
```

Huh. This and MathZoom.js both fail like this.

**Comment by pradyunsg** [comment]
It's not the file size that's the issue here.

MathZoom.js contains the string "</head>". This triggers [`LiveScriptInjector.transform_first_chunk`](https://github.com/lepture/python-livereload/blob/f14bd7e1685e7b8ee1906eb391d924e1f46d26d4/livereload/server.py#L110) in python-livereload, which modifies the header but the chunk isn't supposed to be modified. This breaks an assumption in a different part of the codebase, resulting in this failure.

This is 100% a bug in livereload, one that is not going to be easy to monkeypatch either. >.<

**Comment by pradyunsg** [comment]
> This is 100% a bug in livereload, one that is not going to be easy to monkeypatch either. >.<

Okay, I lied. See #91.

**Comment by tkhyn** [comment]
Thanks a lot, that works provided that `PosixPath` is replaced by `PurePosixPath` so that it can run on Windows too.

It looks like there is a bit of recent activity on python-livereload, maybe submitting a PR other there rather than monkey-patching could fix it for good for other livereload users?

**Comment by pradyunsg** [comment]
> Thanks a lot, that works provided that `PosixPath` is replaced by `PurePosixPath` so that it can run on Windows too.

Pretty sure it'll work, but that's a good catch!

> It looks like there is a bit of recent activity on python-livereload, maybe submitting a PR other there rather than monkey-patching could fix it for good for other livereload users?

I'd like to -- I'm juggling a lot of OSS projects though, so this is gonna sit on the back burner for now.

**Comment by pradyunsg** [comment]
> Pretty sure it'll work, but that's a good catch!

Addressed in c372700c628b706bc330c9f0e3062ef5f15c23c0
